Src/FluentAssertions/Equivalency/DataRelationEquivalencyStep.cs

```diff
@@ -0,0 +1,202 @@
+                if (selectedMembers.TryGetValue(nameof(expectation.ExtendedProperties), out var expectationMember))
+                {
+                    var matchingMember = FindMatchFor(expectationMember, context, config);
+
+                    if (matchingMember != null)
+                    {
+                        IEquivalencyValidationContext nestedContext =
+                            context.CreateForNestedMember(expectationMember, matchingMember);
+
+                        if (nestedContext != null)
+                        {
+                            parent.AssertEqualityUsing(nestedContext);
+                        }
+                    }
+                }
+
+                // Relation constraints
+                CompareConstraints(
+                    parent, context, config, subject, expectation, selectedMembers,
+                    "Child",
+                    selectedMembers.ContainsKey(nameof(DataRelation.ChildTable)),
+                    selectedMembers.ContainsKey(nameof(DataRelation.ChildColumns)),
+                    selectedMembers.ContainsKey(nameof(DataRelation.ChildKeyConstraint)),
+                    r => r.ChildColumns,
+                    r => r.ChildTable);
+
+                CompareConstraints(
+                    parent, context, config, subject, expectation, selectedMembers,
+                    "Parent",
+                    selectedMembers.ContainsKey(nameof(DataRelation.ParentTable)),
+                    selectedMembers.ContainsKey(nameof(DataRelation.ParentColumns)),
+                    selectedMembers.ContainsKey(nameof(DataRelation.ParentKeyConstraint)),
+                    r => r.ParentColumns,
+                    r => r.ParentTable);
+            }
+
+            return true;
+        }
+
+        private static void CompareConstraints(
+            IEquivalencyValidator parent, IEquivalencyValidationContext context, IEquivalencyAssertionOptions config, DataRelation subject, DataRelation expectation, Dictionary<string, SelectedMemberInfo> selectedMembers,
+            string relationDirection,
+            bool compareTable, bool compareColumns, bool compareKeyConstraint,
+            Func<DataRelation, DataColumn[]> getColumns,
+            Func<DataRelation, DataTable> getOtherTable)
+        {
+            if (compareColumns)
+            {
+                var subjectColumns = getColumns(subject);
+                var expectationColumns = getColumns(expectation);
+
+                // These column references are in different tables in different data sets that _should_ be equivalent
+                // to one another.
+                AssertionScope.Current
+                    .ForCondition(subjectColumns.Length == expectationColumns.Length)
+                    .FailWith($"Expected relation referencing {expectationColumns.Length} column(s), but found one referencing {subjectColumns.Length}");
+
+                if (subjectColumns.Length == expectationColumns.Length)
```

Comment:
🤔 And what if the lenghts differ?

Reply:
Then we don't try to compare them. The preceding statement adds a failure message for this same condition.